### ISSUE TITLE
JIT is not exception safe.

### ISSUE THREAD

**Issue by gaika** [arrayfire/arrayfire#1700]
I'm mucking with a Julia wrapper for ArrayFire, and I get these weird dumps any time over ~10 million elements are in a vector.  Same thing would work correctly with a 1-2 million elements `s` vector. Any idea what I'm doing wrong? Underneath it is just calling `af_sum_all` over `af_mul` with batch flag on.
````
julia> afInfo()
ArrayFire v3.5.0 (CUDA, 64-bit Linux, build 1251007)
Platform: CUDA Toolkit 8, Driver: 375.26
[0] GeForce GTX 1080, 8114 MB, CUDA Compute 6.1
-1- Quadro K1200, 4042 MB, CUDA Compute 5.0

julia> typeof(s)
ArrayFire.AFArray{Float32,1}
julia> sum(0.5f0 * s * s)
Only PHI nodes may reference their own value!
  %val1 = fmul float %val0, %val1
Broken module found, compilation terminated!
4115, 1
LOG:

target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:1
28:128-n16:32:64"                                                                                                                    


target triple = "nvptx64-unknown-cuda"

define void @KER8644411623765721933 (
float* %in0,
i32 %dim00,i32 %dim10,i32 %dim20,i32 %dim30,
i32 %str10,i32 %str20,i32 %str30,
float* %out2,

i32 %ostr0, i32 %ostr1, i32 %ostr2, i32 %ostr3,
i32 %odim0, i32 %odim1, i32 %odim2, i32 %odim3,
i32 %blkx, i32 %blky, i32 %ndims )

{

entry:


%tidx = call i32 @llvm.nvvm.read.ptx.sreg.tid.x()
%bdmx = call i32 @llvm.nvvm.read.ptx.sreg.ntid.x()
%bidx = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.x()
%bidy = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.y()
%gdmx = call i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()



%tidy = call i32 @llvm.nvvm.read.ptx.sreg.tid.y()
%bdmy = call i32 @llvm.nvvm.read.ptx.sreg.ntid.y()
%blk_x = alloca i32, align 4
%blk_y = alloca i32, align 4
%id_3 = alloca i32, align 4
%id_2 = alloca i32, align 4
store i32 %bidx, i32* %blk_x, align 4
store i32 %bidy, i32* %blk_y, align 4
store i32 0, i32* %id_2, align 4
store i32 0, i32* %id_3, align 4
%two = alloca i32, align 4
store i32 2, i32* %two, align 4
%twoval = load i32* %two, align 4
%is34 = icmp sgt i32 %ndims, %twoval
br i1 %is34, label %do34, label %do2

do34:
%id2t = sdiv i32 %bidx, %blkx
store i32 %id2t, i32* %id_2, align 4
%id2m = mul i32 %id2t, %blkx
%blk_xx = sub i32 %bidx, %id2m
store i32 %blk_xx, i32* %blk_x, align 4
%three = alloca i32, align 4
store i32 3, i32* %three, align 4
%threeval = load i32* %three, align 4
%is4 = icmp sgt i32 %ndims, %threeval
br i1 %is4, label %do4, label %do2

do4:
%id3t = sdiv i32 %bidy, %blky
store i32 %id3t, i32* %id_3, align 4
%id3m = mul i32 %id3t, %blky
%blk_yy = sub i32 %bidy, %id3m
store i32 %blk_yy, i32* %blk_y, align 4
br label %do2

do2:
%id2 = load i32* %id_2, align 4
%id3 = load i32* %id_3, align 4
%tmp_x = load i32* %blk_x, align 4
%id0m = mul i32 %tmp_x, %bdmx
%id0 = add i32 %tidx, %id0m
%tmp_y = load i32* %blk_y, align 4
%id1m = mul i32 %tmp_y, %bdmy
%id1 = add i32 %tidy, %id1m


%off3o = mul i32 %id3, %ostr3
%off2o = mul i32 %id2, %ostr2
%off1o = mul i32 %id1, %ostr1
%off23o = add i32 %off3o, %off2o
%off123o = add i32 %off23o, %off1o
%idxa = add i32 %off123o, %id0
%idx = sext i32 %idxa to i64


%cmp3 = icmp slt i32 %id3, %odim3
%cmp2 = icmp slt i32 %id2, %odim2
%cmp1 = icmp slt i32 %id1, %odim1
%cmp0 = icmp slt i32 %id0, %odim0
br i1 %cmp3, label %check2, label %end

check2:
br i1 %cmp2, label %check1, label %end

check1:
br i1 %cmp1, label %check0, label %end

check0:
br i1 %cmp0, label %core, label %end

end:

ret void

core:

%b30 = icmp slt i32 %id3, %dim30
%b20 = icmp slt i32 %id2, %dim20
%b10 = icmp slt i32 %id1, %dim10
%b00 = icmp slt i32 %id0, %dim00
%c30 = zext i1 %b30 to i32
%c20 = zext i1 %b20 to i32
%c10 = zext i1 %b10 to i32
%c00 = zext i1 %b00 to i32
%d30 = mul i32 %c30, %id3
%d20 = mul i32 %c20, %id2
%d10 = mul i32 %c10, %id1
%d00 = mul i32 %c00, %id0
%off3i0 = mul i32 %d30, %str30
%off2i0 = mul i32 %d20, %str20
%off1i0 = mul i32 %d10, %str10
%off23i0 = add i32 %off2i0, %off3i0
%off123i0 = add i32 %off23i0, %off1i0
%idxa0 = add i32 %off123i0, %d00
%idx0 = sext i32 %idxa0 to i64

%inIdx0 = getelementptr inbounds float* %in0, i64 %idx0
%val0 = load float* %inIdx0

%val1 = fmul float %val0, %val1
%val2 = fmul float %val1, %val1
%outIdx2= getelementptr inbounds float* %out2, i64 %idx
store float %val2, float* %outIdx2


ret void


}

declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.nctaid.x() nounwind readnone

!nvvm.annotations = !{!1}
!1 = metadata !{void (
float*,
i32, i32, i32, i32,
i32, i32, i32,
float*,
i32, i32, i32, i32,
i32, i32, i32, i32,
i32, i32, i32
)* @KER8644411623765721933,
 metadata !"kernel", i32 1}
ERROR: ArrayFire Error (998) : Internal error
```

**Comment by pavanky** [comment]
What is the size of `s`?

**Comment by gaika** [comment]
```
julia> size(s)
(10000000,)
```

**Comment by gaika** [comment]
here's another one if it is any help:

```
julia> typeof(a)
ArrayFire.AFArray{Int64,1}
julia> size(a)
(1,)
julia> size(s)
(10000000,)
julia> af_sum_all(af_mul(a, sb, true))
3361, 80
LOG:
generated kernel (103, 0): parse instruction forward referenced with type 'i64'
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"


target triple = "nvptx64-unknown-cuda"

define void @KER1392371587324400636 (
float* %out1,

i32 %ostr0, i32 %ostr1, i32 %ostr2, i32 %ostr3,
i32 %odim0, i32 %odim1, i32 %odim2, i32 %odim3,
i32 %blkx, i32 %blky, i32 %ndims )

{

entry:


%tidx = call i32 @llvm.nvvm.read.ptx.sreg.tid.x()
%bdmx = call i32 @llvm.nvvm.read.ptx.sreg.ntid.x()
%bidx = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.x()
%bidy = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.y()
%gdmx = call i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()



%tidy = call i32 @llvm.nvvm.read.ptx.sreg.tid.y()
%bdmy = call i32 @llvm.nvvm.read.ptx.sreg.ntid.y()
%blk_x = alloca i32, align 4
%blk_y = alloca i32, align 4
%id_3 = alloca i32, align 4
%id_2 = alloca i32, align 4
store i32 %bidx, i32* %blk_x, align 4
store i32 %bidy, i32* %blk_y, align 4
store i32 0, i32* %id_2, align 4
store i32 0, i32* %id_3, align 4
%two = alloca i32, align 4
store i32 2, i32* %two, align 4
%twoval = load i32* %two, align 4
%is34 = icmp sgt i32 %ndims, %twoval
br i1 %is34, label %do34, label %do2

do34:
%id2t = sdiv i32 %bidx, %blkx
store i32 %id2t, i32* %id_2, align 4
%id2m = mul i32 %id2t, %blkx
%blk_xx = sub i32 %bidx, %id2m
store i32 %blk_xx, i32* %blk_x, align 4
%three = alloca i32, align 4
store i32 3, i32* %three, align 4
%threeval = load i32* %three, align 4
%is4 = icmp sgt i32 %ndims, %threeval
br i1 %is4, label %do4, label %do2

do4:
%id3t = sdiv i32 %bidy, %blky
store i32 %id3t, i32* %id_3, align 4
%id3m = mul i32 %id3t, %blky
%blk_yy = sub i32 %bidy, %id3m
store i32 %blk_yy, i32* %blk_y, align 4
br label %do2

do2:
%id2 = load i32* %id_2, align 4
%id3 = load i32* %id_3, align 4
%tmp_x = load i32* %blk_x, align 4
%id0m = mul i32 %tmp_x, %bdmx
%id0 = add i32 %tidx, %id0m
%tmp_y = load i32* %blk_y, align 4
%id1m = mul i32 %tmp_y, %bdmy
%id1 = add i32 %tidy, %id1m


%off3o = mul i32 %id3, %ostr3
%off2o = mul i32 %id2, %ostr2
%off1o = mul i32 %id1, %ostr1
%off23o = add i32 %off3o, %off2o
%off123o = add i32 %off23o, %off1o
%idxa = add i32 %off123o, %id0
%idx = sext i32 %idxa to i64


%cmp3 = icmp slt i32 %id3, %odim3
%cmp2 = icmp slt i32 %id2, %odim2
%cmp1 = icmp slt i32 %id1, %odim1
%cmp0 = icmp slt i32 %id0, %odim0
br i1 %cmp3, label %check2, label %end

check2:
br i1 %cmp2, label %check1, label %end

check1:
br i1 %cmp1, label %check0, label %end

check0:
br i1 %cmp0, label %core, label %end

end:

ret void

core:

%val0 = call float @_Z6___mkSx(i64 %val0)
%val1 = fmul float %val0, %val2
%outIdx1= getelementptr inbounds float* %out1, i64 %idx
store float %val1, float* %outIdx1


ret void


}
declare float @_Z6___mkSx(i64)


declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.nctaid.x() nounwind readnone

!nvvm.annotations = !{!1}
!1 = metadata !{void (
float*,
i32, i32, i32, i32,
i32, i32, i32, i32,
i32, i32, i32
)* @KER1392371587324400636,
 metadata !"kernel", i32 1}
ERROR: ArrayFire Error (998) : Internal error
```

**Comment by pavanky** [comment]
Can you start julia with `AF_PRINT_ERRORS=1 julia` and try running the code again ?

**Comment by gaika** [comment]
```
julia> s = fill(1f0, 10_000_000); sa = AFArray(s); sum(mul(a, sa, true))
In function void cuda::evalNodes(std::vector<cuda::Param<T> >&, std::vector<cuda::JIT::Node*>) [with T = float]
In file src/backend/cuda/jit.cpp:658
CU Error (1)


ERROR: ArrayFire Error (998) : Internal error
```

**Comment by gaika** [comment]
Some magic value at 2_097_121
```
julia> for k = 1:1000; a = AFArray([1f0]); sa = AFArray(fill(1f0, 2097100+k)); @show size(sa); sum(mul(a, sa, true)); end
size(sa) = (2097101,)
size(sa) = (2097102,)
size(sa) = (2097103,)
size(sa) = (2097104,)
size(sa) = (2097105,)
size(sa) = (2097106,)
size(sa) = (2097107,)
size(sa) = (2097108,)
size(sa) = (2097109,)
size(sa) = (2097110,)
size(sa) = (2097111,)
size(sa) = (2097112,)
size(sa) = (2097113,)
size(sa) = (2097114,)
size(sa) = (2097115,)
size(sa) = (2097116,)
size(sa) = (2097117,)
size(sa) = (2097118,)
size(sa) = (2097119,)
size(sa) = (2097120,)
size(sa) = (2097121,)
In function void cuda::evalNodes(std::vector<cuda::Param<T> >&, std::vector<cuda::JIT::Node*>) [with T = float]
In file src/backend/cuda/jit.cpp:658
CU Error (1)


ERROR: ArrayFire Error (998) : Internal error
```

**Comment by pavanky** [comment]
@ranjanan I think the instead of printing the output of `af_error_to_string`, you could also add the output of `af_get_last_error` to the error message. It'll be useful for debugging.

@gaika I think the issue is only happening when you are trying to replicate broadcast at larger sizes.
This is because of the issue described in https://github.com/arrayfire/arrayfire/issues/133

Because you are creating an af_constant of size 1 from a scalar, and then doing math operations with another array, the JIT currently thinks the operation needs to be in non linear format and launches a 2D grid that has too few rows and too many columns.  

A work around that might work is to create the 1x1 array using [constant](https://github.com/JuliaComputing/ArrayFire.jl/blob/db52ed6fd0dde102dd09adfde75040b138992a68/src/create.jl#L34) instead of creating an array using something like `AFArray([1f0]);` This is because constant arrays are treated differently than arrays from buffers and result in a different JIT tree which can perform other optimizations.

**Comment by gaika** [comment]
Thanks! It is definitely broadcast related, I'll try it with af_constant next

**Comment by ranjanan** [comment]
I got these errors when using `af_const`: 
```
In function void cuda::evalNodes(std::vector<cuda::Param<T> >&, std::vector<cuda::JIT::Node*>) [with T = double]
In file src/backend/cuda/jit.cpp:658
CU Error (1)
```

**Comment by ranjanan** [comment]
Another one: 
```
3369, 80
LOG:
generated kernel (104, 0): parse instruction forward referenced with type 'i64'
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"


target triple = "nvptx64-unknown-cuda"

define void @KER8123415386589550224 (
double* %out1,

i32 %ostr0, i32 %ostr1, i32 %ostr2, i32 %ostr3,
i32 %odim0, i32 %odim1, i32 %odim2, i32 %odim3,
i32 %blkx, i32 %blky, i32 %ndims )

{

entry:


%tidx = call i32 @llvm.nvvm.read.ptx.sreg.tid.x()
%bdmx = call i32 @llvm.nvvm.read.ptx.sreg.ntid.x()
%bidx = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.x()
%bidy = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.y()
%gdmx = call i32 @llvm.nvvm.read.ptx.sreg.nctaid.x()



%tidy = call i32 @llvm.nvvm.read.ptx.sreg.tid.y()
%bdmy = call i32 @llvm.nvvm.read.ptx.sreg.ntid.y()
%blk_x = alloca i32, align 4
%blk_y = alloca i32, align 4
%id_3 = alloca i32, align 4
%id_2 = alloca i32, align 4
store i32 %bidx, i32* %blk_x, align 4
store i32 %bidy, i32* %blk_y, align 4
store i32 0, i32* %id_2, align 4
store i32 0, i32* %id_3, align 4
%two = alloca i32, align 4
store i32 2, i32* %two, align 4
%twoval = load i32* %two, align 4
%is34 = icmp sgt i32 %ndims, %twoval
br i1 %is34, label %do34, label %do2

do34:
%id2t = sdiv i32 %bidx, %blkx
store i32 %id2t, i32* %id_2, align 4
%id2m = mul i32 %id2t, %blkx
%blk_xx = sub i32 %bidx, %id2m
store i32 %blk_xx, i32* %blk_x, align 4
%three = alloca i32, align 4
store i32 3, i32* %three, align 4
%threeval = load i32* %three, align 4
%is4 = icmp sgt i32 %ndims, %threeval
br i1 %is4, label %do4, label %do2

do4:
%id3t = sdiv i32 %bidy, %blky
store i32 %id3t, i32* %id_3, align 4
%id3m = mul i32 %id3t, %blky
%blk_yy = sub i32 %bidy, %id3m
store i32 %blk_yy, i32* %blk_y, align 4
br label %do2

do2:
%id2 = load i32* %id_2, align 4
%id3 = load i32* %id_3, align 4
%tmp_x = load i32* %blk_x, align 4
%id0m = mul i32 %tmp_x, %bdmx
%id0 = add i32 %tidx, %id0m
%tmp_y = load i32* %blk_y, align 4
%id1m = mul i32 %tmp_y, %bdmy
%id1 = add i32 %tidy, %id1m


%off3o = mul i32 %id3, %ostr3
%off2o = mul i32 %id2, %ostr2
%off1o = mul i32 %id1, %ostr1
%off23o = add i32 %off3o, %off2o
%off123o = add i32 %off23o, %off1o
%idxa = add i32 %off123o, %id0
%idx = sext i32 %idxa to i64


%cmp3 = icmp slt i32 %id3, %odim3
%cmp2 = icmp slt i32 %id2, %odim2
%cmp1 = icmp slt i32 %id1, %odim1
%cmp0 = icmp slt i32 %id0, %odim0
br i1 %cmp3, label %check2, label %end

check2:
br i1 %cmp2, label %check1, label %end

check1:
br i1 %cmp1, label %check0, label %end

check0:
br i1 %cmp0, label %core, label %end

end:

ret void

core:

%val0 = call double @_Z6___mkDx(i64 %val1)
%val1 = fsub double %val0, %val0
%outIdx1= getelementptr inbounds double* %out1, i64 %idx
store double %val1, double* %outIdx1


ret void


}
declare double @_Z6___mkDx(i64)


declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.x() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.y() nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.nctaid.x() nounwind readnone

!nvvm.annotations = !{!1}
!1 = metadata !{void (
double*,
i32, i32, i32, i32,
i32, i32, i32, i32,
i32, i32, i32
)* @KER8123415386589550224,
 metadata !"kernel", i32 1}
In function char* cuda::irToPtx(std::string, size_t*)
In file src/backend/cuda/jit.cpp:415
NVVM Error (9): Failed to compile program
```

**Comment by pavanky** [comment]
That is a weird looking kernel. Can you show me what code generated this ?

**Comment by ranjanan** [comment]
```
julia> using ArrayFire

julia> a = rand(AFArray{Float64}, 1024000, 1, 3);

julia> c = constant(1, 1, 1, 3)
1×1×3 ArrayFire.AFArray{Int64,3}:
[:, :, 1] =
 1

[:, :, 2] =
 1

[:, :, 3] =
 1

julia> a - c
```

**Comment by pavanky** [comment]
ok I am able to reproduce it if I put an `af_eval(c);` between `af_constant` and `af_sub`.  Since you are using the repl, it is evaluating and printing the value which then behaves like a regular buffer.. If you can try the same code from the script, it'll behave differently..

As for your errors, 
- The first error that is showing up is related to issue #133 as mentioned earlier.
- The second error is happening because the JIT is not cleaning up nicely if there is an error. 

Since everything before CUDA 7 is no longer supported for arrayfire 3.5, I am going to go ahead and drop nvvm and move to nvrtc which should make fixing #133 very easy. 

I am changing the scope of the issue to deal with second error instead of marking this issue as duplicate.

**Comment by pavanky** [comment]
Fixed in hotfixes-3.5.1
